libsignal-protocol/src/store/identity_key_store.rs

```diff
@@ -1,16 +1,31 @@
-use crate::{errors::InternalError, Address, Buffer};
 use std::os::raw::{c_int, c_void};
 
-pub trait IdentityKeyStore {
-    // Get the local client's identity key pair.
-    fn identity_key_pair(&self) -> Result<(Buffer, Buffer), InternalError>;
+use crate::{
+    errors::InternalError, keys::IdentityKeyPair, Address, Serializable,
+};
+
+// Provides an interface to identity information.
+pub trait IdentityKeyStore<'a> {
+    /// Get the local client's identity key pair.
+    fn identity_key_pair(&self) -> Result<IdentityKeyPair, InternalError>;
 
     /// Get the local client's registration ID.
     ///
     /// Clients should maintain a registration ID, a random number
     /// between 1 and 16380 that's generated once at install time.
     fn local_registration_id(&self) -> Result<u32, InternalError>;
 
+    /// Save a remote client's identity key
+    ///
+    /// Store a remote client's identity key as trusted.
+    ///
+    /// it should return `true` if the identity key replaces a previous
+    /// identity, `false` if not
+    fn save_identity(
+        &self,
+        address: Address<'a>,
```

Comment:
`rustc` keeps complaining about that life time, as it was `Address<'_>`, see the impl ...

Reply:
We don't want to add `'a` to the trait impl here. That'd mean we're saying all addresses passed to the `IdentityKeyStore` must outlive it. Considering the `IdentityKeyStore` is created at the very start of an application during setup, that means it'd be impossible to load an address at runtime and let C pass it to us (safely).

I think `'_` is the correct way to do things. That says our `Address` lives for the duration of the `save_identity()` call, but can't be used after returning (e.g. imagine it's allocated in the caller's stack frame).

Reply:
Yeah, i need to tackle these life-time issues, feel free to push your solutions to solve that :)